File: frontend/src/lib/api/node.ts
import { requestJson, type ApiRequestOptions } from "./client";
import type {
  JsonObject,
  MessageResponse,
  PaginatedResponse,
  PaginationQuery,
} from "./common";
import type { NodeNetworkResponse } from "./node-network";

export type NodeStatus = "pending" | "approved" | "suspended" | "revoked";

export type AddNodeRequest = {
  address?: number;
  description?: string;
  device_id: string;
  name: string;
  network_id?: string;
  preferences?: JsonObject;
};

export type SetNodeInfoRequest = {
  description?: string;
  name?: string;
};

export type SetNodeNetworkAssignmentRequest = {
  address?: number | null;
  network_id?: string | null;
};

export type SetNodeStatusRequest = {
  status: NodeStatus;
};

export type AddRangingRecordByAddressesRequest = {
  destination_address: number;
  distance: number;
  pan_id: number;
  reported_by_device_id: string;
  source_address: number;
};

export type NodeResponse = {
  address: number | null;
  approved_at: string | null;
  approved_by: string | null;
  created_at: string;
  description: string;
  device_id: string;
  id: string;
  is_approved: boolean;
  last_connected_at: string | null;
  last_disconnected_at: string | null;
  last_seen_at: string | null;
  name: string;
  network: NodeNetworkResponse | null;
  preferences: JsonObject;
  status: NodeStatus;
  updated_at: string | null;
};

export type NodesResponse = PaginatedResponse<NodeResponse>;

export type GetNodesQuery = PaginationQuery & {
  address?: number;
  is_online?: boolean;
  network_id?: string;
  status?: NodeStatus;
};

export type NodeRegistrationResponse = {
  device_id: string;
  is_registered: boolean;
};

export type RegisteredNodesResponse = {
  data: string[];
};

export function createNode(
  request: AddNodeRequest,
  options?: ApiRequestOptions,
): Promise<NodeResponse> {
  return requestJson<NodeResponse>("/nodes", {
    ...options,
    json: request,
    method: "POST",
  });
}

export function getNodes(
  query: GetNodesQuery = {},
  options?: ApiRequestOptions,
): Promise<NodesResponse> {
  return requestJson<NodesResponse>("/nodes", {
    ...options,
    query,
  });
}

export function getRegisteredNodes(
  options?: ApiRequestOptions,
): Promise<RegisteredNodesResponse> {
  return requestJson<RegisteredNodesResponse>("/nodes/registered", options);
}

export function getNodeRegistration(
  deviceId: string,
  options?: ApiRequestOptions,
): Promise<NodeRegistrationResponse> {
  return requestJson<NodeRegistrationResponse>(
    `/nodes/registered/${encodeURIComponent(deviceId)}`,
    options,
  );
}

export function getNodeByDeviceId(
  deviceId: string,
  options?: ApiRequestOptions,
): Promise<NodeResponse> {
  return requestJson<NodeResponse>(
    `/nodes/device/${encodeURIComponent(deviceId)}`,
    options,
  );
}

export function restartNode(
  deviceId: string,
  options?: ApiRequestOptions,
): Promise<MessageResponse> {
  return requestJson<MessageResponse>(
    `/nodes/device/${encodeURIComponent(deviceId)}/restart`,
    {
      ...options,
      method: "POST",
    },
  );
}

export function getNodeByNetworkAddress(
  networkId: string,
  address: number,
  options?: ApiRequestOptions,
): Promise<NodeResponse> {
  return requestJson<NodeResponse>(
    `/nodes/network/${encodeURIComponent(networkId)}/address/${encodeURIComponent(String(address))}`,
    options,
  );
}

export function addRangingRecordByAddresses(
  request: AddRangingRecordByAddressesRequest,
  options?: ApiRequestOptions,
): Promise<MessageResponse> {
  return requestJson<MessageResponse>("/nodes/ranging-records", {
    ...options,
    json: request,
    method: "POST",
  });
}

export function getNode(
  nodeId: string,
  options?: ApiRequestOptions,
): Promise<NodeResponse> {
  return requestJson<NodeResponse>(`/nodes/${encodeURIComponent(nodeId)}`, options);
}

export function updateNodeInfo(
  nodeId: string,
  request: SetNodeInfoRequest,
  options?: ApiRequestOptions,
): Promise<NodeResponse> {
  return requestJson<NodeResponse>(`/nodes/${encodeURIComponent(nodeId)}/info`, {
    ...options,
    json: request,
    method: "PATCH",
  });
}

export function updateNodeNetworkAssignment(
  nodeId: string,
  request: SetNodeNetworkAssignmentRequest,
  options?: ApiRequestOptions,
): Promise<NodeResponse> {
  return requestJson<NodeResponse>(
    `/nodes/${encodeURIComponent(nodeId)}/network`,
    {
      ...options,
      json: request,
      method: "PATCH",
    },
  );
}

export function updateNodePreferences(
  nodeId: string,
  preferences: JsonObject,
  options?: ApiRequestOptions,
): Promise<NodeResponse> {
  return requestJson<NodeResponse>(
    `/nodes/${encodeURIComponent(nodeId)}/preferences`,
    {
      ...options,
      json: preferences,
      method: "PATCH",
    },
  );
}

export function updateNodeStatus(
  nodeId: string,
  request: SetNodeStatusRequest,
  options?: ApiRequestOptions,
): Promise<NodeResponse> {
  return requestJson<NodeResponse>(
    `/nodes/${encodeURIComponent(nodeId)}/status`,
    {
      ...options,
      json: request,
      method: "PATCH",
    },
  );
}

export function deleteNode(
  nodeId: string,
  options?: ApiRequestOptions,
): Promise<MessageResponse> {
  return requestJson<MessageResponse>(`/nodes/${encodeURIComponent(nodeId)}`, {
    ...options,
    method: "DELETE",
  });
}

File: backend/src/ips_app/domain/usecases/node.py
from abc import ABC, abstractmethod
from typing import Any, Dict, List, Optional, Tuple

from ips_app.domain.models.node import Node, NodeStatus


class NodeUsecase(ABC):
    @abstractmethod
    async def create_node(
        self,
        device_id: str,
        name: str,
        description: str = "",
        network_id: Optional[Any] = None,
        address: Optional[int] = None,
        preferences: Optional[Dict[str, Any]] = None,
        created_by: Optional[Any] = None,
    ) -> Node: ...

    @abstractmethod
    async def get_node_by_id(self, id: Any) -> Node: ...

    @abstractmethod
    async def get_node_by_device_id(self, device_id: str) -> Node: ...

    @abstractmethod
    async def get_node_by_network_address(
        self,
        network_id: Any,
        address: int,
    ) -> Node: ...

    @abstractmethod
    async def get_nodes(
        self,
        page: int,
        limit: int,
        search: Optional[str] = None,
        status: Optional[NodeStatus] = None,
        network_id: Optional[Any] = None,
        address: Optional[int] = None,
        is_online: Optional[bool] = None,
    ) -> Tuple[List[Node], int]: ...

    @abstractmethod
    async def update_node_info(
        self,
        id: Any,
        name: Optional[str] = None,
        description: Optional[str] = None,
        updated_by: Optional[Any] = None,
    ) -> Node: ...

    @abstractmethod
    async def update_node_network_assignment(
        self,
        id: Any,
        network_id: Optional[Any],
        address: Optional[int],
        updated_by: Optional[Any] = None,
    ) -> Node: ...

    @abstractmethod
    async def update_node_status(
        self,
        id: Any,
        status: NodeStatus,
        updated_by: Optional[Any] = None,
    ) -> Node: ...

    @abstractmethod
    async def update_node_preferences(
        self,
        id: Any,
        preferences: Dict[str, Any],
        updated_by: Optional[Any] = None,
    ) -> Node: ...

    @abstractmethod
    async def delete_node(self, id: Any) -> None: ...
